code/__HELPERS/AStar.dm

```diff
@@ -61,117 +72,144 @@ Actual Adjacent procs :
 		if(call(start, dist)(end) > maxnodes)
 			return 0
 		maxnodedepth = maxnodes //no need to consider path longer than maxnodes
-
 	var/datum/Heap/open = new /datum/Heap(/proc/HeapPathWeightCompare) //the open list
-	var/list/closed = new() //the closed list
+	var/list/openc = new() //open list for node check
+//	var/list/close = new() //to spare nodes from GC
 	var/list/path = null //the returned path, if any
-	var/datum/PathNode/cur //current processed turf
-
 	//initialization
-	open.Insert(new /datum/PathNode(start,null,0,call(start,dist)(end),0))
-
+	var/datum/PathNode/cur = new /datum/PathNode(start,null,0,call(start,dist)(end),0,15,1)//current processed turf
+	var/datum/PathNode/last = cur
+	var/ne
+	open.Insert(cur)
+	openc[start] = cur
 	//then run the main loop
 	while(!open.IsEmpty() && !path)
-		//get the lower f node on the open list
 		cur = open.Pop() //get the lower f turf in the open list
-		closed.Add(cur.source) //and tell we've processed it
+		//get the lower f node on the open list
 
+		if(ne)
+			last = openc[openc[openc.len]]//list search by key is still costly.Let's avoid it whenever it possible.So if no new open nodes, so be it.
+		if(openc.len>1)
+			last.i = cur.i
+			openc.Swap(cur.i,openc.len)
+		//close += cur
+		openc.len -= 1//we need to serch only in open list.So let's keep turf list tidy
+		ne = TRUE
 		//if we only want to get near the target, check if we're close enough
 		var/closeenough
 		if(mintargetdist)
 			closeenough = call(cur.source,dist)(end) <= mintargetdist
-
 		//if too many steps, abandon that path
 		if(maxnodedepth && (cur.nt > maxnodedepth))
 			continue
-
 		//found the target turf (or close enough), let's create the path to it
 		if(cur.source == end || closeenough)
 			path = new()
 			path.Add(cur.source)
-
 			while(cur.prevNode)
 				cur = cur.prevNode
 				path.Add(cur.source)
-
 			break
-
 		//get adjacents turfs using the adjacent proc, checking for access with id
-		var/list/L = call(cur.source,adjacent)(caller,id, simulated_only)
-		for(var/turf/T in L)
-			if(T == exclude || (T in closed))
-				continue
-
-			var/newg = cur.g + call(cur.source,dist)(T)
-
-			var/datum/PathNode/P = pnodelist[T]
-			if(!P)
-			 //is not already in open list, so add it
-				var/datum/PathNode/newnode = new /datum/PathNode(T,cur,newg,call(T,dist)(end),cur.nt+1)
-				open.Insert(newnode)
-				pnodelist[T] = newnode
-			else //is already in open list, check if it's a better way from the current turf
-				if(newg < P.g)
-					P.prevNode = cur
-					P.g = (newg * L.len / 9)
-					P.calc_f()
-					P.nt = cur.nt + 1
-					open.ReSort(P)//reorder the changed element in the list
+		for(var/i = 0 to 3)
+			var/f= 1<<i
+			if(cur.bf & f)
+				var/T = get_step(cur.source,f)
+				if(T != exclude)
+					var/datum/PathNode/CN = openc[T]  //current checking turf
+					var/r=((f & 85)<<1)|((f & 170)>>1) //fuck list calls. bit operations faster
```

Comment:
you need to document the shit out of this

add comments everywhere explaining what/why clearly.

it is acceptable to have wtf levels of code for performance but we need to leave plenty of comments to explain what each step is doing and how the bit hackery is working.

IN a case like this I would recommend going over the top rather than under

Reply:
it's just getting reverse direction throught swapping even and odd bits.

Reply:
```
var/r=((f & 85)<<1)|((f & 170)>>1) //Get the reverse direction by swapping even and odd bits.
```
It needs to be understandable *in code* so when somebody else starts working on the code they can understand everything.

Reply:
plus it will help me review and understand your change, which improves the chances of it being merged

Reply:
also this criticism does not apply to only this line, I was pointing out the most obvious instance and asking you to address all of them in your change.

Anywhere you're doing bitflipping you should ideally document why/what.